explore/matrix-io/02.sparse-and-dense.ipynb

```diff
@@ -0,0 +1,427 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "from scipy import sparse\n",
+    "import sys\n",
+    "import os\n",
+    "import time"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "np.random.seed(0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Sparse .npz (compressed) takes the least disk space regardless of density\n",
+    "Uncompressed sparse matrices take more disk space when the matrices are very dense"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CCpGeAeGaD (1391, 137)\n",
+      "density: 97.8 %\n",
+      "dense-uncompressed: 1.525 MB\n",
+      "sparse-compressed: 1.347 MB\n",
+      "sparse-uncompressed: 1.347 MB\n",
+      "\n",
+      "DaGpBP (137, 11381)\n",
+      "density: 21.0 %\n",
+      "dense-uncompressed: 12.47 MB\n",
+      "sparse-compressed: 2.529 MB\n",
+      "sparse-uncompressed: 2.529 MB\n",
+      "\n",
+      "MFpGdCcSE (2884, 5734)\n",
+      "density: 14.1 %\n",
+      "dense-uncompressed: 132.3 MB\n",
+      "sparse-compressed: 17.79 MB\n",
+      "sparse-uncompressed: 17.79 MB\n",
+      "\n",
+      "GiGpBP (20945, 11381)\n",
+      "density: 4.65 %\n",
+      "dense-uncompressed: 1.907e+03 MB\n",
+      "sparse-compressed: 73.81 MB\n",
+      "sparse-uncompressed: 73.81 MB\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "for i in ['CCpGeAeGaD', 'DaGpBP', 'MFpGdCcSE', 'GiGpBP']:\n",
+    "    mat = np.load(f'data/{i}-dense-uncompressed.npy')\n",
+    "    print(i, mat.shape)\n",
+    "    \n",
+    "    density = 100 * (mat != 0).sum() / np.prod(mat.shape)\n",
+    "    print(f'density: {density :.3} %')\n",
+    "\n",
+    "    sparse_mat = sparse.csc_matrix(mat, copy=True)\n",
+    "\n",
+    "    sparse.save_npz(f'data/{i}-sparse-compressed.npz', sparse_mat, compressed=True)\n",
+    "    sparse.save_npz(f'data/{i}-sparse-uncompressed.npz', sparse_mat, compressed=False)\n",
+    "    \n",
+    "    dense_size = os.path.getsize(f'data/{i}-dense-uncompressed.npy') / 1000000\n",
+    "    sparse_size = os.path.getsize(f'data/{i}-sparse-compressed.npz') / 1000000\n",
+    "    sparse_un_size = os.path.getsize(f'data/{i}-sparse-compressed.npz') / 1000000\n",
+    "    print(f'dense-uncompressed: {dense_size :.4} MB\\nsparse-compressed: {sparse_size :.4} MB\\nsparse-uncompressed: {sparse_un_size :.4} MB\\n')"
```

Comment:
Let's change formatter to `.4g`, which should do 4 significant digits and will not do `1.908e+03`.